infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuInterClusterReplicationServer.java

```diff
@@ -290,21 +304,84 @@ private ServerContext getServerContext(Map<String, Object> opts) {
     }
 
     /**
-     * Start Corfu Log Replication Discovery Service
+     * Start Corfu Log Replication Discovery Service after the local node version matches the cluster version, i.e.,
+     * any ongoing rolling upgrade completes.
      *
      * @param serverContext server context (server information)
      */
     private void startDiscoveryService(ServerContext serverContext) {
 
+        CorfuRuntime runtime = getRuntime(serverContext);
+        CorfuStore corfuStore = new CorfuStore(runtime);
+
+        LogReplicationUpgradeManager upgradeManager =
+                new LogReplicationUpgradeManager(corfuStore, serverContext.getPluginConfigFilePath());
+
+        // Check if an upgrade is in progress.  If it is, wait for it to complete
+        log.info("Wait for any ongoing rolling upgrade to complete.");
+
+        while(isUpgradeInProgress(upgradeManager, corfuStore)) {
+            // Wait before checking again.
+            sleep();
+        }
+
+        // Upgrade has completed.  Start the discovery Service
         log.info("Start Discovery Service.");
 
         // Start LogReplicationDiscovery Service, responsible for
         // acquiring lock, retrieving Site Manager Info and processing this info
         // so this node is initialized as Source (sender) or Sink (receiver)
-        replicationDiscoveryService = new CorfuReplicationDiscoveryService(serverContext);
+        replicationDiscoveryService = new CorfuReplicationDiscoveryService(serverContext, runtime);
         replicationDiscoveryService.start();
     }
 
+    private CorfuRuntime getRuntime(ServerContext serverContext) {
+
+        String localCorfuEndpoint = getCorfuEndpoint(getHostFromEndpointURL(serverContext.getLocalEndpoint()),
```

Comment:
There will be a conflict but init of ClusterManager should not be brought here.  It will continue to get instantiated when DiscoveryService starts, like before.